### ISSUE TITLE
Explain how autocommit works

### ISSUE THREAD

**Issue by xaprb** [VividCortex/go-database-sql-tutorial#33]
What if the database server is set to autocommit=0? Then every new connection is in a transaction automatically. Seems like this will cause problems. We should explain whether this is true or not.

**Comment by arnehormann** [comment]
Not targeting the issue but the fact... **yuck** :-1: 
Did you write a test to check what it does in practice?

**Comment by xaprb** [comment]
Not yet - I intend to. If you want to, go ahead :)

**Comment by arnehormann** [comment]
That honor shall be yours...
Maybe it's time for a Go issue to create something sessiony in database/sql - `db.Session()` with `Close()` or `End()` which has a subset of the functions of `sql.DB`: Query, QueryRow, Exec, Prepare, Begin, Close.
That would also be practical with `SET @var := ...` for variables that have to be kept for multiple queries.

**Comment by xaprb** [comment]
Do you still think this could be a problem? I haven't heard of it, and I wonder if we do something sensible in the driver (you know the code, I don't) to set the autocommit flag _on_ when creating statements that are not part of a transaction. Or is this just left up to the winds of fate?

**Comment by arnehormann** [comment]
We do not check any server flags except max_packet_size and what the protocol delivers in the headers. I'd have to check and do some reading, but I don't intend to do that for now.

Intuitively, I'd say it is a problem but either everyone with that setting knows what causes the error and writes code accordingly or - more probable - nobody uses that setting. I can't think of a sensible reason to use that mode, it probably breaks a lot of tools and libraries and _enforcing_ usage of transactions for every query - the only value it delivers - will be damaging in at least some cases for most real world systems.

**Comment by xaprb** [comment]
I believe, but I'm not sure, that most libraries want to assume autocommit
is set on by default, and that you'll set it off if you want -- during
opening a connection most likely. I am pretty sure this is how Perl's DBI
wants it to be... yep. https://metacpan.org/pod/DBI#AutoCommit "Drivers
should always default to AutoCommit mode (an unfortunate choice largely
forced on the DBI by ODBC and JDBC conventions.)"

I remember once early in my career we configured the server with
autocommit=0 but we figured out our mistake pretty fast.
